services/app/frontend/Dockerfile

```diff
@@ -0,0 +1,40 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+#
+FROM apache/apisix:2.15.0-debian
+
+ARG SW_AGENT_NODEJS_FRONTEND_VERSION
+
+RUN apt update -y && apt install nodejs npm git -y 
+
+WORKDIR /skywalking-client-js
+
+RUN git clone --recurse-submodules https://github.com/apache/skywalking-client-js.git . ; \
+    git reset --hard ${SW_AGENT_NODEJS_FRONTEND_VERSION} ; \
+    git submodule update ; \
+    git submodule sync ; \
+    npm install ; \
+    npm run build
+    
+WORKDIR /app
+
+COPY . .
```

Comment:
We can make multi stage build so we don't need to install packages and save the build time. Like

```Dockerfile
FROM node:<tag> as build

# build the front end 

FROM apisix:<tag>

COPY --from=build <the build folder> <target folder>

```

Reply:
Could  I  use  apt  to install  git ?  Copying git is hard , because is has many lib dependency .

Reply:
And needn't to execute apt update before  installing  git

Reply:
I really understand `multi stage build` , I think . I have updated the docker file . 
Thanks for  your help.